README.md
XWifi


CHANGELOG
Version 1.4

Bug Fixes/Improvements
	Script1 bug in time conversion resolved
	Logger third page final submit wait until element_to_be_clickable instead
		of presence, elementnotinteractable exception issue
	Logger third page military time conversion no longer adds 12 to 12am
		creating value of 24, rather a value of 0
	Logger, on caught exceptions that do not have a special case, now logs
		timestamp and more info on nature of error to firefox.log. Func
		portal_authenticate loops over each page for execution.
	Logger no longer saves copys firefox temp cache, only closes webdriver
		
Notes
	- Xfinity updated UI to use 3 pages instead of 4, no more zip code 
		aswell
	- Logger currently suffers from inability to be headless or minimization
		of browser window. Explicit wait for element_to_be_clickable is the 
		likely culprit.
	- timeout set to wait 20s and long_wait 25 as default
	- media.cache_readahead_limit in firefox about:config could be adjusted 
		for smoother video watching between cycles

--------------------------------------------------------------------------------

Version 1.3

Bug Fixes/Improvements
	Staple checks and finds firefox binary directory, propogates directory
		info to Script1 -> (Script2, Logger) 
	Staple now makes sure that DIG_DNS can resolve after checking network-manager
		is a running process, instead of doing check in Script1
	Script1 does not check to shred txt, or log files
	Script1 logs stderr and stdout to firefox.log
	Script1 improved logic for controling python execution and error controll
	Logger has log_path in webdriver for generated geckodriver.log to /dev/null
	Logger waits to get password fields before stoping web javascript
	Staple, Script1 and Script2 uses NIC that not just BROADCAST but also UP
	
Notes
	- If more than one NIC exits, first one to be found by command ip, will be
		used. For specific device, ifconfig "interface" down all others
	- Launching firefox always creates empty log file, there maybe a better
		way than to check if empty => delete everytime fireox is launnched

--------------------------------------------------------------------------------

Version 1.2

Bug Fixes/Improvements
	Script1 fixed logger retrying form fillout actions
	Script1 removed default time
	Script1 using arrays to handle exit and retry_attempts
	Script1 Restarts cycle if no free option found thus trying new mac, else
		retry attempts for any other reason relaunches firefox
	Logger had no set_time operation
	Logger fixed parameter value assignment when retry attempt
	Logger fixed timeout and driver_timeout global variable update
	Logger Has 3 failed states that could be caught; Failed to launch,
		No free Option page 1, Failed to fill of 1st,2nd,3rd or 4th page 
Notes
	- No long term testing, seems stable
	- Still clunky DIG_DNS resolution, its possible that the worst resolution
		path is not reachable 
	- Hard-coded string in logger for firefox binary directory should be made
		to be dynamic
		
--------------------------------------------------------------------------------

Version 1.1

Bug Fixes/Improvements
	Staple, when executing, follows symlink instead of relying on locate
		and correct naming structure
	When checking if Selenium profile exists, instead of relative path uses 
		absolute path
	Symlink created from $PARENT_DIR not hardcoded string
	Removed lines for redundant check for global_vars file and creation,
		firefox profile prefs.js pre removal, reference to undefined var
		$STARTING_DIR
	Script1 reduced time to wait to start pinging xfinity to 1 second
	Script1 fixed stale DIG_DNS, update every 6 failed attemps
	Script1 dig option +noedns added. Ignores server malformed packets 
	Script1 creates lock file to manage Script2 monitoring
	Script1	$T_ACT value never exceeds $T_REF value during normal cycle
	Script1 and Script2 use same command to identify which interface to use,
		no total shutdown due to mismatched when 2 or more network 
		interfaces present
	Script2 no more nm dependence, checks /sys/class/net/ directory
	Script2 curl query max time 2 seconds
	Script2 uses Script1 lock file instead of cumbersome strace
	Script2 now checks and resolves no default gateway
	Script2 tries to enable default device if down, else exits
	Script2 Checks every 3 seconds, less sleeping if required to ReSync Script1
	Logger new Class and Function structures. Smart useragent, email, fake name,
		generation, and collision avoidance. 
		
Notes
	- Script1 logger.py launch retry on error is broken
	- A minuscule chance that logger will reach a collision if macchanger sets 
	  a previously known by Comcast mac addr
	- DIG_DNS can have a clunky path to resolution
	- Unconfirmed if saving firefox cache fills up free disk space
	- Need more testing to determine efficacy of Script2 checks and 
	  resolutions
	- 2 or more NICs may cause unforseen problems
	  
--------------------------------------------------------------------------------

Version 1.0

New Features
	Refer dhcp to use preset Comcast DNS, if possible
	Subvert to 3rd party DNS After Comcast Authentication  
	Staple Defined in Path
	Use/Create Custom Firefox Profile, maintain Comcast Cache
	Force temp Selenium profile to make permenant changes   
	
Bug Fixes/Improvements
	Ensure requested dig is translated, if ! xxx.xxx.xxx.xxx -> try again
	No garbage temp files outside target directory*
	Removed redundant locating of target directory in disowned children
	Script2 no dig loop fixed
	If logger timesout, relaunches but already has Auth, set 30 min as default 
	Kill only logger firefox children
	Remove firefox cookies with specificity 

Notes
 -  staple has harcoded string looking for  ~/.mozilla/firefox and passing to
	logger.py
 -  logger.py has hardcoded strings for firefox-esr binary directory, receives
	string passed from staple, and expects ~/.cache/mozilla/firefox to be 
	a valid directory
 -  network-manager and nmcli is require to run staple

script1.sh
#!/bin/bash

CYAN='\033[0;36m'
RED='\033[0;31m'
YELLOW='\033[1;33m'
NC='\033[0m'

CYCLECOUNT=0
SLOWPING=6
RETRY_ATTEMPTS=(2 2) #1st is actual value, 2nd same value to replace 1st 

IFACE=$( ip -br link | awk '/<BROADCAST,MULTICAST,UP/ {print $1; exit}' )
RES_DNS=$( cat /etc/resolv.conf 2> /dev/null )
DIG_DNS=$( dig +time=3 +tries=1 +noedns +noall +answer +short \
		  xfwweb.g.comcast.net @75.75.75.75 )

function check_firefoxlog_empty {
	[[ -s './firefox.log' ]] || shred -fzu -n 1 './firefox.log' 
}
		  
#Check That Firefox Profile, and Bin Dir was Passed
if [ -z $1 ] ; then
	echo -e "[${RED}!${NC}] Error with Firefox Profile [${RED}!${NC}]\n"
	echo -e "\tEXITING SCRIPT\n"
	sleep 7
	exit
elif [ -z $2 ] ; then
	echo -e "\n[${RED}!${NC}] Error with Firefox Bin Dir [${RED}!${NC}]\n"
	echo -e "\tEXITING SCRIPT\n"
	sleep 7
	exit	
fi 

SELENIUM_FIREFOX="$1"
SELENIUM_F_BIN="$2"

#Garbage Collection
shred -fzu -n 1 *.out *.lock wget* 2>/dev/null &

while : ; do
	clear
	#Prevent Script2 from interfering
	touch script1.lock
	
	#Network Card Reset Block
	systemctl stop network-manager.service
	ifconfig $IFACE down 
	sudo macchanger -r -b $IFACE 
	ifconfig $IFACE up
	systemctl start network-manager.service

	#Store New Values in global_vars 
	macchanger -s $IFACE | awk -F " " '/Current/ {print $3}' > "global_vars.csv";
	echo -e "\nWaiting for Connection [${YELLOW}?${NC}] \n"
	
	sleep 1 # Dont Spam NetworkManager Prematurely 
		
	#Wait For Established Connection	
	while true ; do
		wget --spider --timeout=1 --tries=1 $DIG_DNS  > /dev/null 2>&1
			if [[ $? -eq 0 ]] ; then
				echo -e "Connected [${RED}+${NC}]\n"
				break
			else
				if ! (( $SLOWPING % 6)) ; then
					DIG_DNS=$(dig +time=1 +tries=1 +noedns +noall +answer +short \
								xfwweb.g.comcast.net @75.75.75.75)
					echo -e  "${CYAN}	Pinging Xfinity...${NC}\n"
				fi
				((SLOWPING++))
				sleep 0.75
			fi
	done

	echo -e "Headless Firefox Launched [${YELLOW}+${NC}]\n"
	
	python3 logger.py "$SELENIUM_FIREFOX" "$SELENIUM_F_BIN" >> firefox.log 2>&1 
	EXIT_RETRY=($? 1)
	check_firefoxlog_empty 
	# Catch error code from Firefox, keep retrying for RETRY_ATTEMPTS amount

	if [[ ${EXIT_RETRY[0]} -eq 255 && ${RETRY_ATTEMPTS[0]} -lt 1 ]] ; then 					#No Free Option, No attempts left, exit
		echo -e "[${RED}!${NC}] Err: No Free Opt Dectected, Exiting [${RED}!${NC}]\n"	
		sleep 7
		killall xterm > /dev/null 2>&1
	elif [[ ${EXIT_RETRY[0]} -eq 255 && ${RETRY_ATTEMPTS[0]} -ne 0 ]] ; then				#No Free Option, restart at the begining
		echo -e "[${YELLOW}!${NC}] Err: No Free Opt Dectected, Retrying with New Mac [${YELLOW}!${NC}]\n"	
		((RETRY_ATTEMPTS[0]--))
		continue
	elif [[ ${EXIT_RETRY[0]} -eq 254 && ${RETRY_ATTEMPTS[0]} -lt 1 ]] ; then 				#1 of the 4 Forms Failed to complete, No attempts left, exit
		echo -e "[${RED}!${NC}] Err: Failed To Complete a Form Page [${RED}!${NC}]\n"
		sleep 7
		killall xterm > /dev/null 2>&1
	elif [[ ${EXIT_RETRY[0]} -eq 254 && ${RETRY_ATTEMPTS[0]} -ne 0 ]] ; then				#1 of the 4 Forms Failed to complete
		echo -e "[${YELLOW}!${NC}] Err: Failed to Complete a Form Page [${YELLOW}!${NC}]\n"	
		sleep 2
	elif [[ ${EXIT_RETRY[0]} -ne 0 ]] ; then												#Failed to Launch, exit
		echo -e "[${RED}!${NC}] Err: Failed To Launch Firefox [${RED}!${NC}]\n"	
		sleep 7
		#Clean Up Dead Children
		kill -9 $(ps -ef | grep -P '(?!.*(/bin/bash|xterm -T))^(?=.*pts)(?=.*firefox)' \
						 | awk '{ print $2 }') 2>/dev/null
		killall xterm > /dev/null 2>&1
	fi

	while [[ ${EXIT_RETRY[0]} -ne 0  && ${RETRY_ATTEMPTS[0]} -ne 0 ]] ; do					#Try for RETRY_ATTEMPTS to re Launch
		echo -e "Headless Firefox [${CYAN}Re${NC}]Launched [${YELLOW}!${NC}]\n"
		((EXIT_RETRY[1]++))
		
		python3 logger.py "$SELENIUM_FIREFOX" "$SELENIUM_F_BIN" ${EXIT_RETRY[1]} >> firefox.log 2>&1 
		EXIT_RETRY[0]=$?
		check_firefoxlog_empty

		((RETRY_ATTEMPTS[0]--))
		
		#Clean Up Dead Children
		kill -9 $(ps -ef | grep -P '(?!.*(/bin/bash|xterm -T))^(?=.*pts)(?=.*firefox)' \
						 | awk '{ print $2 }') 2>/dev/null
	done

	if [[ ${EXIT_RETRY[0]} -ne 0 && ${EXIT_RETRY[1]} -ne 0 ]] ; then 						#Tried to Launch without Failure, No attempts left
		echo -e "[${RED}!${NC}] Err: Form Could Not Be Filled Out [${RED}!${NC}]\n"
		sleep 7
		killall xterm >/dev/null 2>&1
		exit
	fi

	#Remove Prescribed DNS and Replace With 3rd Party DNS 
	if [ -n "$RES_DNS" ] ; then
		cat > /etc/resolv.conf <<- EOM 
			nameserver 8.8.8.8
			nameserver 1.1.1.1
			nameserver 8.8.4.4
			nameserver 1.0.0.1
		EOM
	else
		echo -e "\nResolv File Missing or No DNS values Were Generated[${RED}!${NC}]"
		echo -e "\n${YELLOW}CHECK${NC} '/etc/resolv.conf' ${YELLOW}FILE${NC}\n"
		sleep 3
	fi
	
	shred -fzu -n 1 'script1.lock'
	
	#Check If Script2 Is Already Running
	if [ -z "$(pgrep script2.sh)" ] ; then
		sudo xterm -T "Xfinity Wifi" -geometry 70x5+0-0 -fa monospace -fs 8 \
				-e "./script2.sh $SELENIUM_FIREFOX $SELENIUM_F_BIN" & disown
	fi; clear
	
	((CYCLECOUNT++))
	echo -e "Entering Current Cycle at [${CYAN}$(date +%I:%M ) $(date +%p)${NC}]\n"
	echo -e "Cycle Number: [${YELLOW}$CYCLECOUNT${NC}]\n"

	#Check if Default is Set, Then Reference Time Given by Xfinity
	TIME=($(awk -F ',' 'NR==2 { gsub(/\r/,"",$0); print $1, $2 }' global_vars.csv))
	
	HOURS=$(( ${TIME[0]} - ($(date +%-H) % 23) ))
	OFFSET=20 #Margin of Error in seconds 
	
	#Determine Duration, deriving 12H Time from Military Time
	if [ ${TIME[0]} -eq 0 ] ; then 						#23H -> 00H
		MINS=$(( ( 60 - $(date +%-M)) + ${TIME[1]} ))
	elif [ $HOURS -gt 0 ] ; then						#[1H-22H] -> [2H-23H]
		HOURS=$(( $HOURS - 1 ))
		MINS=$(( ${TIME[1]} + (60 - $(date +%-M)) ))	
	else 
		MINS=$(( ${TIME[1]} - $(date +%-M) ))			# >1H, Mins Only
	fi

	SECS=$(( ($HOURS * 600) + ($MINS * 60) - $OFFSET))
	T_REF=$(date +%s -d +${SECS}sec)
	
	while [ $SECS -gt 1 ] ; do
		T_ACT=$(( ($(date +%s) - 1) - $OFFSET))
		#Hibernating Device ReSync
		if [ $T_REF -lt $T_ACT ] ; then
			echo -e "SysTime is Out Of Sync Due to Hibernating Device [${RED}!${NC}]"
			sleep 1.5
			SECS=0
		fi
		echo -ne "Countdown Until Next Cycle [${YELLOW}$SECS${NC}]\033[0K\r"
		sleep 1
		: $((SECS--))
	done
	RETRY_ATTEMPTS[0]=${RETRY_ATTEMPTS[1]} #Restore Retry Attempts to Previous Value
done


script2.sh
#!/bin/bash

RED='\033[0;31m'
YELLOW='\033[1;33m'
NC='\033[0m'
CYAN='\033[0;36m'

DOUBLE_CHECK=1
COUNTER=7
SLEEP_TIME=3

#Check That Firefox Profile, and Bin Dir was Passed
if [ -z $1 ] ; then
	echo -e "[${RED}!${NC}] Error with Firefox Profile [${RED}!${NC}]\n"
	echo -e "\tEXITING SCRIPT\n"
	sleep 7
	exit
elif [ -z $2 ] ; then
	echo -e "\n[${RED}!${NC}] Error with Firefox Bin Dir [${RED}!${NC}]\n"
	echo -e "\tEXITING SCRIPT\n"
	sleep 7
	exit	
fi 

SELENIUM_FIREFOX="$1"
SELENIUM_F_BIN="$2"

DEFAULT_DEVICE=$( ip -br link | awk '/<BROADCAST,MULTICAST,UP/ {print $1; exit}' )

#Main Loop, Ensure Connectivity
while : ; do
	clear
	echo -e "Active$CYAN FIDELITY$NC Monitoring [${RED}+${NC}] \n"
	sleep $SLEEP_TIME
	
	DIG_DNS=$( dig +time=1 +tries=1 +noall +noedns +answer +short google.com )
	VALUE=$( wpa_cli -i $DEFAULT_DEVICE scan_results | grep -m 1 "xfinitywifi" )
   	AVAIL=$( curl -I -k -m 2 -s -w %{http_code} -o /dev/null $DIG_DNS )
   	WIFI_ID=$( iwgetid -r )
   	
   	if [[ ! -f 'script1.lock' ]] ; then  
		STATE=$( cat /sys/class/net/$DEFAULT_DEVICE/operstate )
		ONLINE=$( cat /sys/class/net/$DEFAULT_DEVICE/carrier )
		
		#Check if DEFAULT_DEVICE is up
		if [[ "$STATE" == "down" ]] ; then
			clear
			echo -e "[${YELLOW}!${NC}] Err: DEFAULT DEVICE IS NOT UP [${YELLOW}!${NC}]\n"
			sleep 2
			ifconfig $DEFAULT_DEVICE up
			if [ $? -eq 0 ] ; then
				continue;
			else 
				echo -e "\n[${RED}!${NC}] Err: IFCONFIG FAILED TO RAISE DEFAULT DEVICE [${RED}!${NC}]\n"
				sleep 7
				killall xterm >/dev/null 2>&1
				exit
			fi
		fi
		
		#Check if DEFAULT_DEVICE is online
		if [[ -z "$VALUE" && "$ONLINE" -eq 1 ]] ; then
			clear
			echo -e "Xfinitywifi Is No Longer In The Area \n"
			echo -e	"$CYAN	   Goodbye$NC [$YELLOW!$NC] \n"
			sleep 7
			killall xterm >/dev/null 2>&1
			exit
		fi
		
		#Check if bad default gateway
		DEFAULT_GW=$( ip route show )
		if [[ -z $DEFAULT_GW ]] ; then
			clear		
			systemctl restart network-manager 
			echo -e "[${YELLOW}!${NC}] Err: NO DEFAULT GATEWAY [${YELLOW}!${NC}]\n"
			sleep 2
			while true ; do 
				ip route show
				if [[ $? -eq 0 ]] ; then
					sleep 10
					break 
				fi
				sleep 0.5
			done
			continue
		fi
		
		#If Connected To Xfinity, Check If There is Internet Access
		if [[ "$AVAIL" == "000" && "$WIFI_ID" == "xfinitywifi" ]] ; then
			if [[ `expr $DOUBLE_CHECK % 5` -eq 0 ]]; then
				DOUBLE_CHECK=1
				SCRIPT_CHECK=$(pgrep -f script1.sh)
				echo -e "Internet$YELLOW DISCONTINUITY$NC Detected [${YELLOW}!${NC}] \n"
				sleep 1
				clear
				if [ -n "$SCRIPT_CHECK" ]; then
					for pid in $(pgrep -f script1); do kill -9 $pid; done
				fi
				sudo xterm -T "Xfinity Wifi" -geometry 70x20+0+0 -fa monospace -fs 8 \
						-e "./script1.sh $SELENIUM_FIREFOX $SELENIUM_F_BIN" & disown
				echo -e "Active$CYAN FIDELITY$NC Monitoring$YELLOW ReSynced$NC [${YELLOW}+${NC}]"
				sleep 2
				COUNTER=7
				continue
			fi
			((DOUBLE_CHECK++))
			sleep 1
		fi
	fi
	if [ "$DOUBLE_CHECK" -gt 1 ]; then
		((COUNTER--))
		SLEEP_TIME=2
		if [[ "$COUNTER" -eq 0 && "$DOUBLE_CHECK" -ne 1 && `expr $DOUBLE_CHECK % 5` -ne 0 ]]; then 
			DOUBLE_CHECK=1
			COUNTER=7
			SLEEP_TIME=10
		fi
	fi	
done


firecache/user.js
user_pref("browser.dom.window.dump.enabled", true);
user_pref("browser.laterrun.enabled", false);
user_pref("browser.newtab.url", "about:blank");
user_pref("browser.newtabpage.enabled", false);
user_pref("browser.startup.page", 0);
user_pref("browser.startup.homepage", "about:blank");
user_pref("browser.startup.homepage_override.mstone", "ignore");
user_pref("browser.usedOnWindows10.introURL", "about:blank");
user_pref("dom.max_chrome_script_run_time", 30);
user_pref("dom.max_script_run_time", 30);
user_pref("dom.report_all_js_exceptions", true);
user_pref("javascript.options.showInConsole", true);
user_pref("network.captive-portal-service.enabled", false);
user_pref("security.csp.enable", false);
user_pref("startup.homepage_welcome_url", "about:blank");
user_pref("startup.homepage_welcome_url.additional", "about:blank");
user_pref("webdriver_accept_untrusted_certs", true);
user_pref("webdriver_assume_untrusted_issuer", true);
user_pref("webdriver_enable_native_events", true);
user_pref("app.update.auto", false);
user_pref("app.update.enabled", false);
user_pref("browser.displayedE10SNotice", 4);
user_pref("browser.download.manager.showWhenStarting", false);
user_pref("browser.EULA.override", true);
user_pref("browser.EULA.3.accepted", true);
user_pref("browser.link.open_external", 2);
user_pref("browser.link.open_newwindow", 2);
user_pref("browser.offline", false);
user_pref("browser.reader.detectedFirstArticle", true);
user_pref("browser.safebrowsing.enabled", false);
user_pref("browser.safebrowsing.malware.enabled", false);
user_pref("browser.search.update", false);
user_pref("browser.selfsupport.url", "");
user_pref("browser.sessionstore.resume_from_crash", false);
user_pref("browser.shell.checkDefaultBrowser", false);
user_pref("browser.tabs.warnOnClose", false);
user_pref("browser.tabs.warnOnOpen", false);
user_pref("datareporting.healthreport.service.enabled", false);
user_pref("datareporting.healthreport.uploadEnabled", false);
user_pref("datareporting.healthreport.service.firstRun", false);
user_pref("datareporting.healthreport.logging.consoleEnabled", false);
user_pref("datareporting.policy.dataSubmissionEnabled", false);
user_pref("datareporting.policy.dataSubmissionPolicyAccepted", false);
user_pref("devtools.errorconsole.enabled", true);
user_pref("dom.disable_open_during_load", false);
user_pref("extensions.autoDisableScopes", 10);
user_pref("extensions.blocklist.enabled", false);
user_pref("extensions.checkCompatibility.nightly", false);
user_pref("extensions.update.enabled", false);
user_pref("extensions.update.notifyUser", false);
user_pref("javascript.enabled", true);
user_pref("network.manage-offline-status", false);
user_pref("network.http.phishy-userpass-length", 255);
user_pref("offline-apps.allow_by_default", true);
user_pref("prompts.tab_modal.enabled", false);
user_pref("security.fileuri.origin_policy", 3);
user_pref("security.fileuri.strict_origin_policy", false);
user_pref("signon.rememberSignons", false);
user_pref("toolkit.networkmanager.disable", true);
user_pref("toolkit.telemetry.prompted", 2);
user_pref("toolkit.telemetry.enabled", false);
user_pref("toolkit.telemetry.rejected", true);
user_pref("xpinstall.signatures.required", false);
user_pref("xpinstall.whitelist.required", false);


logger.py
import datetime, time, secrets, platform, os, sys, set_time, csv, shutil, re, inspect

from faker import Faker
from hashlib import blake2b
from selenium import webdriver
from python_arptable import ARPTABLE
from user_agent import generate_user_agent
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.firefox.firefox_binary import FirefoxBinary
from selenium.webdriver.support import expected_conditions as EC

from selenium.common.exceptions import ( TimeoutException,
	NoSuchElementException, StaleElementReferenceException )

_random = secrets.SystemRandom()
randrange = _random.randrange

timeout_multiplier = 1
driver_timeout = 15 # <-- Adjust in seconds
driver_long_timeout = 25

server_gettime_req = (
	f'''var date = new Date($.ajax({{async: false, type: 'GET', '''
	f'''contentType: 'application/json;charset=utf-8' '''
	f'''}}).getResponseHeader( 'Date' )); return [ '''
	f'''date.getFullYear(), date.getMonth() + 1, date.getDate(), '''
	f'''date.getHours(), date.getMinutes(), date.getSeconds(), 0] '''
)

def check_exists_by_xpath( xpath ):
	try:
		driver.find_element_by_xpath( xpath )
	except NoSuchElementException:
		return False
	return True
	
def check_input( parameters ):
	profile_dir = bin_dir = None
	global driver_timeout
	global driver_long_timeout
	global timeout_multiplier
	
	if len(parameters) == 3:
		profile_dir = str(sys.argv[1])
		bin_dir = str(sys.argv[2])
	
	elif len(sys.argv) == 4:
		try:
			profile_dir = str(sys.argv[1])
			bin_dir = str(sys.argv[2])
			timeout_multiplier = int(sys.argv[3])
		except ValueError:
			print("Integer Value as Third Parameter only!")
			timeout_multiplier = 3
			
	driver_timeout = timeout_multiplier * driver_timeout
	driver_long_timeout = timeout_multiplier * driver_long_timeout
	
	return profile_dir, bin_dir
	
def initialize_profile( f_profile_dir ):				
	firefox_profile = webdriver.FirefoxProfile( f_profile_dir )
	return firefox_profile
		
def initialize_driver( f_profile_dir, f_binary, f_profile, ua ):
	sel_slice1, sel_slice2 = f_profile_dir.split("/.mozilla", 1)
	f_cache_dir = sel_slice1 + "/.cache/mozilla" + sel_slice2

	f_options = webdriver.FirefoxOptions()
	f_options.headless = True
	f_options.accept_insecure_certs = True
	f_options.set_capability("pageLoadStrategy","none")
	f_options.set_preference("browser.cache.disk.parent_directory",
										f_cache_dir)
	f_options.set_preference("general.useragent.override", 
										ua )
	driver = webdriver.Firefox( firefox_binary=f_binary, options=f_options, 
								firefox_profile=f_profile, service_log_path=os.devnull)
	return driver

def portal_authenticate( obj ):

	pages = [ method for method in dir( obj ) if ( method.endswith('page') ) ]
	for page in pages:
		try:
			getattr( obj, page )()
		except:
			driver.quit()
			print(f'{datetime.datetime.now()} - {page} Error {{\n{sys.exc_info()} }}\n')
			sys.exit(254)

class wait_for_text_to_match( object ):
    def __init__(self, locator, pattern):
        self.locator = locator
        self.pattern = re.compile(pattern)

    def __call__(self, driver):
        try:
            element_text = EC._find_element(driver, self.locator).text
            return self.pattern.search(element_text)
        except StaleElementReferenceException:
            return False

class User:
	__slots__ = ['first_name','last_name']
        
	def __init__( self, first_name, last_name ):
		self.first_name = first_name
		self.last_name = last_name
		
	salt = bytes( '%.4f' % time.process_time_ns(), 'utf8' )

	def salted_hash( self, digest_size, message ):
		haesh = blake2b( digest_size=digest_size, salt=self.salt )
		haesh.update(b'message')

		return haesh.hexdigest()

	@property
	def email( self ):    
		message = f'''{self.first_name}.{self.last_name}'''
		username = self.salted_hash( 6 , message )
		email_string = ( f'''{username}@comcast.com''' )

		return email_string
	
	@property
	def password( self ):
		message = f'''{self.last_name}.{self.first_name}'''
		password = self.salted_hash( 4 , message )
		return password
		
	@property
	def unq_hostname( self ):
		hostname = secrets.token_urlsafe(randrange(5,8))
		return hostname
        
	@property
	def url_wifiondemand( self ):
		with open ("global_vars.csv", mode = 'r') as global_vars:
			spoofed_mac = global_vars.read().replace(":","%3A").strip()
		ap_mac = ARPTABLE[0]["HW address"].replace(":","%3A").strip()
		user_os = platform.system()
		hostname = self.unq_hostname
		
		url = ( f'''https://wifiondemand.xfinity.com/wod/landing?c=n&macId='''
			f'''{spoofed_mac}&a=as&bn=st22&location=default&apMacId={ap_mac}'''
			f'''&issuer=r&deviceModel={user_os}+Chrome+-+&deviceName={hostname}''' )
		return url

class XfinityForms( User ):
	def __init__( self, driver ):
		generator = Faker()
		User.__init__( self, generator.first_name(), generator.last_name() )
		self.driver = driver
		self.wait = WebDriverWait( driver, driver_timeout )						 
		self.long_wait = WebDriverWait( driver, driver_long_timeout )						 
	
	def first_page( self ):
		driver = self.driver
		wait = self.wait
		
		driver.get( self.url_wifiondemand )
		
		try:
			free_opt = wait.until(EC.presence_of_element_located((By.XPATH,"//*[contains(text(), 'Complimentary')]")))
		except TimeoutException:
			driver.quit()
			print("No Free Option To Select")
			sys.exit(255)
				
		first_submit = wait.until(EC.presence_of_element_located((By.ID,"continueButton")))
		driver.execute_script("window.stop();")

		free_opt.click()
		
		time_tuple = driver.execute_script(server_gettime_req)
		set_time._linux_set_time(time_tuple)
		
		first_submit.click()

		if check_exists_by_xpath("/html/body/div[1]/div/div[3]/div[2]"):
			decline = wait.until(EC.presence_of_element_located((By.XPATH,"//button[@id='upgradeOfferCancelButton']")))
			decline.click()

	def second_page( self ):
		driver = self.driver
		wait = self.wait
		unq_hostname = self.unq_hostname
			
		first_name_box = wait.until(EC.presence_of_element_located((By.XPATH,"//*[@id='firstName']")))		
		last_name_box = driver.find_element_by_xpath("//*[@id='lastName']")
		
		driver.execute_script("window.stop();")

		first_name_box.send_keys(self.first_name)
		last_name_box.send_keys(self.last_name)
						
		password = driver.find_element_by_xpath("//input[@id='password']")
		password_retype = driver.find_element_by_xpath("//input[@id='passwordRetype']")
		password.send_keys(unq_hostname + "$")
		password_retype.send_keys(unq_hostname + "$")
	
		username = wait.until(EC.element_to_be_clickable((By.XPATH,"//button[@id='usePersonalEmail']"))).click()
		email_box = driver.find_element_by_xpath("//*[@id='primaryEmail']")
		email_box.send_keys(self.email)
	
		drop_menu = wait.until(EC.element_to_be_clickable((By.XPATH,"/html/body/main/div/form/div[4]/fieldset/select/option[2]")))
		secret_answer = driver.find_element_by_xpath("//input[@id='secretAnswer']")	
		drop_menu.click()
		secret_answer.send_keys(unq_hostname)
	
		wait.until(EC.element_to_be_clickable((By.XPATH,"//*[@id='submitButton']"))).click()

	def third_page( self ):
		driver = self.driver
		wait = self.wait
		long_wait = self.long_wait
		
		time_message = long_wait.until(EC.presence_of_element_located((By.XPATH,"//span[@id='orderConfirmationSponsoredExpirationDate']"))).text
				
		driver.execute_script("window.stop();")
	
		time_m_list = time_message.strip().split(' ')	
		final_submit = long_wait.until(EC.element_to_be_clickable((By.XPATH,"//button[@id='_orderConfirmationActivatePass']")))

		military_time = time_m_list[6].split(':')

		#Check Time then Convert to Military Time 
		if (time_m_list[7] == 'PM' and str(military_time[0]) != '12'):
			hour_time = str((int(military_time[0]) + 12))
		elif (time_m_list[7] == 'AM' and str(military_time[0]) == '12'):
			hour_time = "0"
		else:
			hour_time = military_time[0]
			
		min_time = military_time[1].lstrip("0")
		
		try:
			with open ("global_vars.csv", mode='a') as global_vars:
				csv_writer = csv.writer(global_vars)
				csv_writer.writerow([hour_time, min_time])
		finally:
			global_vars.close()
		
		if final_submit.is_enabled():
			final_submit.click()
			
	def tear_down( self, f_profile, f_profile_dir ):
		# src_files = os.listdir(f_profile.path)
		# root_src_dir = f_profile.path
		# root_dst_dir = f_profile_dir

		# for src_dir, dirs, files in os.walk(root_src_dir):
			# dst_dir = src_dir.replace(root_src_dir, root_dst_dir, 1)
			# if not os.path.exists(dst_dir):
				# os.makedirs(dst_dir)
			# for file_ in files:
				# src_file = os.path.join(src_dir, file_)
				# dst_file = os.path.join(dst_dir, file_)
				# if os.path.exists(dst_file):
					# # in case of the src and dst are the same file
					# if os.path.samefile(src_file, dst_file):
						# continue
					# os.remove(dst_file)
				# shutil.move(src_file, dst_dir)
		driver.quit()
		
f_profile_dir, f_binary_dir = check_input( sys.argv )
f_profile = initialize_profile( f_profile_dir )
user_agent = generate_user_agent()

driver = initialize_driver( f_profile_dir, f_binary_dir, f_profile, user_agent )

xfinity = XfinityForms( driver )
portal_authenticate( xfinity )
xfinity.tear_down( f_profile, f_profile_dir )
